feat: change submit method

mathiasbio opened the pull request:
## Description

<!-- Provide a brief overview of your PR and link any relevant user stories or issues. -->

[PR description]

#### Added

- [Description]

#### Changed

- [Description]

#### Fixed

- [Description]

#### Removed

- [Description]

## Documentation

<!-- Link all added or updated documents. -->

- [ ] N/A
- [ ] Updated Balsamic documentation to reflect the changes as needed for this PR.
    - [Document Name]

## Tests

<!-- Describe in detail how you tested your changes to help reviewers validate the code. -->
<!-- Include screenshots or visual representations of your changes. -->

### Feature Tests

<!-- Include tests relevant to the changes in this PR. -->

- [ ] N/A
- [ ] Test [Description]
    - [Screenshot]

### Pipeline Integrity Tests

<!-- Include tests to verify the integrity of the different Balsamic workflows. -->

- **Report deliver (generation of the `.hk` file)**
    - [x] N/A
    - [ ] Verified
- **TGA T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **TGA T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **QC Workflow**
    - [x] N/A
    - [ ] Verified
- **PON Workflow**
    - [x] N/A
    - [ ] Verified

## Clinical Genomics Stockholm

<!-- Do not reveal clinical data, and if applicable, place it within the internal Google Drive directory. -->

### Documentation

<!-- Link related issues or PRs for necessary changes. -->

- **Atlas documentation**
    - [x] N/A
    - [ ] Updated: [Link]
- **Web portal for Clinical Genomics**
    - [x] N/A
    - [ ] Updated: [Link]

### Panel of Normal specific criteria

<!-- If the PR includes a new PoN adhere to the criteria below. -->

- [ ] The PR includes the addition of a new Panel of Normals
- [ ] The samples have been verified to adhere to the sample selection criteria on [Atlas PoN creation instructions for Balsamic](https://atlas.scilifelab.se/infrastructure/BALSAMIC/balsamic_pon/#procedure)

### User Changes

<!-- Please provide justification if you select N/A and the changes affect the output or results. -->

- [x] N/A
- [ ] This PR affects the output files or results.
    - [ ] User feedback is considered unnecessary because [Justification].
    - [ ] Affected users have been included in the development process and given a chance to provide feedback.

### Infrastructure Changes

<!-- Link related issues or PRs for necessary changes. -->
<!-- Make sure that the linked PRs include relevant tests. -->

- **Stored files in Housekeeper**
    - [x] N/A
    - [ ] Updated: [Link]
- **CG (CLI and delivered/uploaded files)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Servers (configuration files on Hasta)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Scout interface**
    - [x] N/A
    - [ ] Updated: [Link]

### Validation criteria

<!-- Create validation criteria for this PR or mark as unnecessary. -->

Validation criteria to be added to validation report PR: [LINK-TO-VALIDATION-REPORT-PR from the [validations repository](https://github.com/Clinical-Genomics/validations/)]

**Version specific criteria**

- Text here or N/A



> [!IMPORTANT]
> One of the below checkboxes for validation need to be checked 

- [ ] Added version specific validation criteria to validation report
- [ ] Changes validated in standard sections: [validation-section]
- [ ] Validation criteria not necessary

## Checklist

> [!IMPORTANT]  
> Ensure that all checkboxes below are ticked before merging.

### For Developers

- **PR Description**
    - [ ] Provided a comprehensive description of the PR.
    - [ ] Linked relevant user stories or issues to the PR.
- **Documentation**
    - [ ] Verified and updated documentation if necessary.
- **Validation criteria**
    - [ ] Completed the validation criteria section of the template.
- **Tests**
    - [ ] Described and tested the functionality addressed in the PR.
    - [ ] Ensured integration of the new code with existing workflows.
    - [ ] Confirmed that meaningful unit tests were added for the changes introduced.
    - [ ] Checked that the PR has successfully passed all relevant code smells and coverage checks.
- **Review**
    - [ ] Addressed and resolved all the feedback provided during the code review process.
    - [ ] Obtained final approval from designated reviewers.

### For Reviewers

- **Code**
    - [ ] Code implements the intended features or fixes the reported issue.
    - [ ] Code follows the project's coding standards and style guide.
- **Documentation**
    - [ ] Pipeline changes are well-documented in the CHANGELOG and relevant documentation.
- **Validation criteria**
    - [ ] The author has completed the validation criteria section of the template
- **Tests**
    - [ ] The author provided a description of their manual testing, including consideration of edge cases and boundary
      conditions where applicable, with satisfactory results.
- **Review**
    - [ ] Confirmed that the developer has addressed all the comments during the code review.

<!-- Add any other relevant information or specific checks necessary for your PR. -->

fevac commented on BALSAMIC/models/snakemake.py:
```diff
@@ -155,27 +136,9 @@ def get_snakemake_cluster_options(self) -> str:
-            "{dependencies}'"
-        )
-        return remove_unnecessary_spaces(cluster_submit_command)
+        return "--default-resources mem_mb=32000 threads=8"
```
why do we even need this?

mathiasbio replied:
Good question 😂 to be honest I primarily added this in order to make our --dry-run pytest work, because the --dry-run is only run locally it doesn't have access to the cluster.yaml and I don't think the workflow profile would work either. And if we don't supply any resources then the --dry-run will crash because snakemake can't find any resources for the rules. So basically...this is just adding some default resources if you're running balsamic locally, which nobody ever does, and I doubt it would work very well... 🤔

mathiasbio replied:
Maybe I can add some comment in the code about it?

fevac commented on BALSAMIC/snakemake_rules/align/tga_bam_postprocess.rule:
```diff
@@ -10,15 +10,15 @@ rule sentieon_dedup_consensus:
         score = Path(bam_dir, "{sample_type}.{sample}.dedup.score").as_posix(),
         dedup_metrics = Path(qc_dir, "{sample_type}.{sample}.dedup_conensus.metrics").as_posix()
+    resources:
+        mem_mb = lambda wc, input: max(3.2 * input.size_mb, 5000)
```
it's nice to have all resources specified in the same location, plus we might want to overwrite that if 5000 would become too restrictive

mathiasbio replied:
The 5GB is just the minimum requested memory. But yeah, I don't know if this is just an issue of me not saving the reasons why I put this or if I was just confused. I'll run some more tests. But the logic behind it would be that we're running panels of such different sizes that it could be nice to not request too much memory for small panels unnecessarily, but now that I compare WES and myeloid panel for instance I don't see much difference in the benchmark for memory...so maybe the amount of memory isn't that sensitive and it's better to just have a static. But I'll run some more tests!

mathiasbio replied:
I ended up removing this and using a fixed mem. It seems to be working fine!

fevac reviewed (commented):
Really nice!! 🌟  The code also became much cleaner!! I just have few comments, and there is some code smell you should also check for security reasons

mathiasbio commented on BALSAMIC/snakemake_rules/align/wgs_bam_postprocess.rule:
```diff
@@ -9,13 +9,13 @@ rule sentieon_dedup:
     benchmark:
         Path(benchmark_dir, "sentieon_dedup_{sample_type}.{sample}.tsv").as_posix()
+    resources:
+        mem_mb = lambda wc, input: min(max(3.2 * input.size_mb, 10000), 500000) # Cap at 500GB
```
I don't remember exactly what it was that made me put the total memory. I seem to remember that there was a case that really struggled to finish with memory, but now I can't find it. Either way, this lamba function would only go up to a maximum of 500GB, usually it would demand much less. But let me re-run some WGS case and test...because at the moment multiplying total size with 3.2 sounds like a bit too much...

mathiasbio commented:
@fevac I added back 1 retry because of the recent fails of VarDict in production. I also removed the dynamic memory assignment and added static mem instead. I am still making sure that a large WGS TN case and a TGA TN + UMI goes through the pipeline fine. But once they can, I would feel good about merging this. There is some decrease in code coverage but it doesn't seem to be relevant to add tests for this according to me 🙏

mathiasbio commented:
3 main items of feedback from the discussion with sys-dev:
- for prod, can we either add the job-fail summaries directly to the standard error of the headjob, or add a link to the file?
- add a new argument to balsamic for snakemake head-job slurm partition
- investigate if the slurm job-status checks can be more standardised, such as with using slurm executor plugin mentioned here: https://github.com/snakemake/snakemake/issues/2477#issuecomment-2092932177  (using this would require upgrading snakemake, which could run into issues that I have forgotten what they were...)

mathiasbio commented on BALSAMIC/assets/scripts/cluster_job_status.py:
```diff
@@ -0,0 +1,70 @@
+jobid = sys.argv[1]
+
+
+def run(cmd, timeout=10):
```
Thanks for the reviews on this script! It has been removed now in favour of the changes done in the downstream PR : https://github.com/Clinical-Genomics/BALSAMIC/pull/1604 so I will resolve the comments involving this script

mathiasbio commented on BALSAMIC/snakemake_rules/align/tga_bam_postprocess.rule:
```diff
@@ -12,13 +12,13 @@ rule sentieon_dedup_consensus:
     benchmark:
         Path(benchmark_dir, "sentieon_dedup_{sample_type}.{sample}.tsv").as_posix()
+    resources:
+        mem_mb = lambda wc, input: max(5.1 * input.size_mb, 5000)
```
For the most part having a static memory not dependent on the input size of the files was working for this rule, but for one particularly high covered TGA case it needed much more memory. Like 80GB or something, which we really don't need for the myeloid. I tried steadily increasing it bit by bit, and this was the multiplying factor that worked. The 5000MB is simply the minimum mem used which I don't expect will ever be reached. But if the input size * 5.1 is ever smaller than 5GB it should be fine, not like we're wasting much resources at that point

mathiasbio commented on BALSAMIC/models/sbatchsubmitter.py:
```diff
@@ -0,0 +1,148 @@
+        self.conda_env_path = os.environ.get("CONDA_PREFIX", "")
+        self.sbatch_script_path = self.script_path / "BALSAMIC_snakemake_submit.sh"
+
+    def create_sbatch_script(self) -> None:
```
This function has been somewhat refactored in the downstream PR maybe it's more readable now

mathiasbio commented on BALSAMIC/commands/config/case.py:
```diff
@@ -129,6 +135,21 @@ def case_config(
+    add_file_logging(log_file, logger_name=__name__)
+
+    LOG.info(f"Running BALSAMIC version {balsamic_version} -- CONFIG CASE")
+    LOG.info(f"BALSAMIC started with log level {context.obj['log_level']}.")
```
Sure, removed!

mathiasbio commented on BALSAMIC/commands/config/case.py:
```diff
@@ -129,6 +135,21 @@ def case_config(
+    LOG.info(f"Running BALSAMIC version {balsamic_version} -- CONFIG CASE")
+    LOG.info(f"BALSAMIC started with log level {context.obj['log_level']}.")
+
+    LOG.info("Collecting reference and annotation file paths.")
```
Sure, removed!

mathiasbio commented on BALSAMIC/commands/config/case.py:
Probably not very useful logs, I removed some

mathiasbio commented on BALSAMIC/commands/options.py:
```diff
@@ -148,25 +139,38 @@
+    help="Directory containing snakemake workflow profile specifying rule resources",
+)
+
+OPTION_CACHE_PROFILE = click.option(
```
This is for the rules involved in the balsamic cache generation, before we had the cluster_cache.json file, now this is in a yaml instead. I think it's kind of nice to have the two split up since the main one for running balsamic is already enormous, and it might be necessary to have some slightly different settings (like maybe avoiding reruns when building cache)

mathiasbio commented on BALSAMIC/assets/scripts/scan_finished_jobid_status.py:
```diff
@@ -0,0 +1,94 @@
+def get_job_state(jobid):
+    try:
+        result = subprocess.run(
+            ["/usr/bin/scontrol", "show", "job", jobid],
```
This has been changed in the downstream PR there we're using just "scontrol"

mathiasbio commented on BALSAMIC/commands/config/case.py:
```diff
@@ -40,7 +40,12 @@
+from BALSAMIC.constants.analysis import (
+    BIOINFO_TOOL_ENV,
+    AnalysisWorkflow,
+    Gender,
```
I'll refactor all the places Gender is used to Sex instead, in the downstream PR

mathiasbio replied:
It's a bit annoying though as we're using this inconsistently in different places, like input argument from CG is gender, not sex...

mathiasbio commented on BALSAMIC/commands/run/analysis.py:
```diff
@@ -66,22 +66,38 @@ def analysis(
+
+    LOG.info(f"Initializing balsamic config model from config JSON: {sample_config}.")
+    sample_config_path: Path = Path(sample_config).absolute()
+    with open(sample_config_path, "r") as sample_fh:
```
I think the inconsistency comes from how the input argument is sample_config, which is a filepath, is the same name as when the config is read into a dict later, but I needed to preserve the sample_config as a path so I added another variable. But maybe instead I can just rename the input argument...but that would also require changing the name in CG and now it feels like too much work

mathiasbio commented:
Thanks for the review @fevac ! I don't really know how to organise this now with the PRs, in some ways I could just abandon this PR and continue with this one: https://github.com/Clinical-Genomics/BALSAMIC/pull/1604 but I kind of like to keep this for traceability. So maybe the upstream one can be reviewed, then merged into this one, and the whole thing then merged into develop?

mathiasbio commented:
Closing this PR as it was replaced by this one https://github.com/Clinical-Genomics/BALSAMIC/pull/1604
